### PR TITLE
[JSON Logic] Part 7: Add conditional logic checking 

### PR DESCRIPTION
**Pull request by brennj**
Definitely a more involved one 😅 : 

- We will parse conditionals now in `x-jsf-logic` via a special `processJSONLogicNode` function which goes through anything particular to JSON logic.
- During the calculation of conditional properties, we'll now calculated computed attributes if any and spread them, we'll also apply any validations that need to be added.
-  Adds the logic for the special if statements only available in `x-jsf-logic` (which can have an if statement without `properties` but `computedValues` and `validations` checks)
- Lots of tests and schemas with more to come (breaking some extra tests into its own PR because theres just so much...)

### CODE REVIEW

File: src/tests/jsonLogic.test.js [hunk 1]
```diff
@@ -322,6 +326,74 @@ describe('cross-value validations', () => {
...
+  describe('Conditionals', () => {
+    it('when field_a > field_b, show field_c', () => {
+      const { fields, handleValidation } = createHeadlessForm(
+        schemaWithGreaterThanChecksForThreeFields,
+        { strictInputType: false }
+      );
+      expect(fields.find((i) => i.name === 'field_c').isVisible).toEqual(false);
+
+      expect(handleValidation({ field_a: 1, field_b: 3 }).formErrors).toEqual(undefined);
+      expect(handleValidation({ field_a: 1, field_b: null }).formErrors).toEqual({
```
[the remark this answers is not in the document]
Comment (sandrina-p) [reply src/tests/jsonLogic.test.js:338]: Uff I think I found a clean solution but please double-check it. Here's the [public Loom](https://www.loom.com/share/d263864013b64303993ebd8409a26f9d)

File: src/jsonLogic.js [hunk 2]
```diff
@@ -249,6 +254,67 @@ function handleNestedObjectForComputedValues(values, formValues, parentID, valid
...
+  }
+
+  if (node.if) {
+    const matchesPropertyCondition = checkIfConditionMatches(
+      node,
+      formValues,
+      formFields,
+      validations
+    );
+    const matchesValidationsAndComputedValues = checkIfMatchesValidationsAndComputedValues(
```
Comment (sandrina-p) [inline src/jsonLogic.js:284]: performance: Shall we run this (evaluate jsonLogic) only if `matchesPropertyCondition` is true?

```suggestion
    const matchesValidationsAndComputedValues = matchesPropertyCondition && checkIfMatchesValidationsAndComputedValues(
```

File: src/tests/jsonLogicFixtures.js [hunk 3]
```diff
@@ -179,204 +198,330 @@ export const schemaWithValidationThatDoesNotExistOnProperty = {
...
-      is_even_number: {
-        errorMessage: 'B must be even',
-        rule: {
-          '===': [{ '%': [{ var: 'field_b' }, 2] }, 0],
+    },
+    allOf: [
+      {
+        if: {
+          validations: {
+            require_c: {
```
Comment (sandrina-p) [inline src/tests/jsonLogicFixtures.js:278]: todo: for readability, i think this should be called `a_bigger_than_b` or not? we need guidelines on how to name these validations 😅  because technically this is not making c required. That's only on the `then` as side effect.

File: src/tests/jsonLogic.test.js [hunk 4]
```diff
@@ -322,6 +326,74 @@ describe('cross-value validations', () => {
...
+      const cField = fields.find((i) => i.name === 'field_c');
+      expect(cField.isVisible).toEqual(false);
+      expect(cField.description).toEqual(undefined);
+      expect(handleValidation({ field_a: 10, field_b: 5 }).formErrors).toEqual({
+        field_c: 'Required field',
+      });
+      expect(handleValidation({ field_a: 10, field_b: 5, field_c: 0 }).formErrors).toEqual({
+        field_c: 'Needs more numbers',
+      });
+      expect(cField.description).toBe('I am a description!');
```
Comment (sandrina-p) [inline src/tests/jsonLogic.test.js:375]: what a joy!!

Comment (sandrina-p) [reply src/tests/jsonLogic.test.js:375]: ah but it's missing the else. it's a edge-case that happened in the past where you trigger `then`, but if you trigger else again, it get's stuck on `then`. Could you add a last one with a: 5; b: 10; and assert the description is null and the field is invisible?

File: src/createHeadlessForm.js [hunk 5]
```diff
@@ -137,15 +137,16 @@ function buildFieldParameters(name, fieldProperties, required = [], config = {})
  */
 function convertJSONSchemaPropertiesToFieldParameters(
   { properties, required, 'x-jsf-order': order },
-  config = {}
+  config = {},
+  validations
```
[the remark this answers is not in the document]
Comment (brennj) [reply src/createHeadlessForm.js]: its removed now and all good here. I want to say this will have a role for the inline rules in the later PR (being a lazy dev and not verified this), let see.

File: src/helpers.js [hunk 6]
```diff
@@ -475,6 +494,9 @@ export function extractParametersFromNode(schemaNode) {
 
   return omitBy(
     {
+      const: node.const,
+      // This is a "forced value" when both const and default are present.
+      ...(node.const && node.default ? { value: node.const } : {}),
```
[the remark this answers is not in the document]
Comment (brennj) [reply src/helpers.js]: this is gone. i'm actually kinda confused how part 2 works without it 🤷

File: src/jsonLogic.js [hunk 7]
```diff
@@ -249,6 +254,67 @@ function handleNestedObjectForComputedValues(values, formValues, parentID, valid
...
+      .map((allOfNode) =>
+        processJSONLogicNode({ node: allOfNode, formValues, formFields, validations, parentID })
+      )
+      .forEach(({ required: allOfItemRequired }) => {
+        allOfItemRequired.forEach(requiredFields.add, requiredFields);
+      });
+  }
+
+  if (node.if) {
+    const matchesPropertyCondition = checkIfConditionMatches(
```
[the remark this answers is not in the document]
Comment (brennj) [reply src/jsonLogic.js]: What do you think of changing `checkIfConditionMatches` -> `checkIfConditionMatchesProperties` to make this clearer? Done in https://github.com/remoteoss/json-schema-form/pull/41/commits/b1e6772da317c19d11aae77b5d3b5db05bd1230a

File: src/jsonLogic.js (hunk 2 above)
Comment (brennj) [reply src/jsonLogic.js]: this is done, lovely suggestion

File: src/jsonLogic.js [hunk 8]
```diff
@@ -249,6 +254,67 @@ function handleNestedObjectForComputedValues(values, formValues, parentID, valid
...
+    const matchesValidationsAndComputedValues = checkIfMatchesValidationsAndComputedValues(
+      node,
+      formValues,
+      validations,
+      parentID
+    );
+
+    const isConditionMatch = matchesPropertyCondition && matchesValidationsAndComputedValues;
+
+    if (isConditionMatch && node.then) {
```
[the remark this answers is not in the document]
Comment (brennj) [reply src/jsonLogic.js]: lovely refactor, added in 6fbb9e1b334f9efaae652a1f92440f75b5e490aa

File: src/calculateConditionalProperties.js [hunk 9]
```diff
@@ -69,14 +70,14 @@ function rebuildFieldset(fields, property) {
...
-export function calculateConditionalProperties(fieldParams, customProperties) {
+export function calculateConditionalProperties(fieldParams, customProperties, validations, config) {
   /**
    * Runs dynamic property calculation on a field based on a conditional that has been calculated
    * @param {Boolean} isRequired - if the field is required
    * @param {Object} conditionBranch - condition branch being applied
    * @returns {Object} updated field parameters
    */
-  return (isRequired, conditionBranch) => {
+  return (isRequired, conditionBranch, __, _, formValues) => {
```
[the remark this answers is not in the document]
Comment (brennj) [reply src/calculateConditionalProperties.js]: these are updated and i also took the chance to update the function signatures to remove those ugly __'s.

File: src/tests/jsonLogic.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (brennj) [reply src/tests/jsonLogic.test.js]: https://github.com/remoteoss/json-schema-form/pull/41/commits/3fd0092385a4d3079cd72d96f054362c9625d3fa :D

File: src/tests/jsonLogic.test.js [hunk 10]
```diff
@@ -322,6 +326,74 @@ describe('cross-value validations', () => {
...
+      );
+      expect(fields.find((i) => i.name === 'field_c').isVisible).toEqual(false);
+
+      expect(handleValidation({ field_a: 1, field_b: 3 }).formErrors).toEqual(undefined);
+      expect(handleValidation({ field_a: 1, field_b: null }).formErrors).toEqual({
+        field_b: 'Required field',
+      });
+      expect(handleValidation({ field_a: 10, field_b: 3 }).formErrors).toEqual({
+        field_c: 'Required field',
+      });
```
[the remark this answers is not in the document]
Comment (brennj) [reply src/tests/jsonLogic.test.js]: ae07dbf9eb0c1137890d6979c3c90e1a692db323 :)

File: src/tests/jsonLogic.test.js (hunk 4 above)
Comment (brennj) [reply src/tests/jsonLogic.test.js]: I can do the isVisible part easy, but I'm finding the description part tricky. It always comes back with `I am a description!` because it doesn't know _where_ to update. I don't think also fields should be wiped if they aren't visible 🤔 

Gonna pick away it a bit more.

File: src/tests/jsonLogicFixtures.js (hunk 3 above)
Comment (brennj) [reply src/tests/jsonLogicFixtures.js]: changed, good point. Some internal docs i guess around working with JSON schema 🤔

**Comment by brennj** [comment]
I think everything is covered here, let me know if I'm missing anything

File: src/calculateConditionalProperties.js [hunk 11]
```diff
@@ -98,17 +99,37 @@ export function calculateConditionalProperties(fieldParams, customProperties) {
...
+        ...calculatedComputedAttributes,
+        ...(calculatedComputedAttributes.value
+          ? { value: calculatedComputedAttributes.value }
+          : { value: undefined }),
+        schema: buildYupSchema(
+          {
+            ...fieldParams,
+            ...restNewFieldParams,
+            ...calculatedComputedAttributes,
+            requiredValidations,
```
[the remark this answers is not in the document]
Comment (sandrina-p) [reply src/calculateConditionalProperties.js]: even better 💯

File: src/helpers.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (sandrina-p) [reply src/helpers.js]: same here, but to be honest, if it works now and it's covered with unit tests, that's what matters. (no time for deep debug session now)

File: src/jsonLogic.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (sandrina-p) [reply src/jsonLogic.js]: the older I get the more I like long variables x)

File: src/tests/jsonLogic.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (sandrina-p) [reply src/tests/jsonLogic.test.js]: _puuuur_ fect 😻

File: src/tests/jsonLogicFixtures.js (hunk 3 above)
Comment (sandrina-p) [reply src/tests/jsonLogicFixtures.js]: I hope you are taking note of these little things for later when we open the PR about json-logic docs ^.^

File: src/tests/jsonLogic.test.js (hunk 4 above)
Comment (sandrina-p) [reply src/tests/jsonLogic.test.js]: I believe that bug is related to [this internal issue](https://linear.app/remote/issue/RMT-58/jsf-conditionals-keywords-are-redundant-eg-readonly-false)... if it is, then leave it as it is (maybe with the assertions commented out and a code comment mentioning the issue id (RMT-58)

File: src/tests/jsonLogic.test.js [hunk 12]
```diff
@@ -363,58 +371,205 @@ describe('jsonLogic: cross-values validations', () => {
...
+
+    it('Use an inline rule for a minimum, maximum value', () => {
+      const { fields, handleValidation } = createHeadlessForm(
+        schemaInlineComputedAttrForMaximumMinimumValues,
+        {
+          strictInputType: false,
+        }
+      );
+      const [, fieldB] = fields;
+      expect(fieldB).toMatchObject({ minimum: NaN, maximum: NaN });
```
Comment (brennj) [inline src/tests/jsonLogic.test.js:404]: @sandrina-p ... can you please review this test specifically based off our convo https://github.com/remoteoss/json-schema-form/pull/40#discussion_r1311418085

is this right for the initial value?

File: src/helpers.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (brennj) [reply src/helpers.js]: FWIW, it ended up needed to be put back in part two so we can rest easy with this one now.

File: src/tests/jsonLogic.test.js (hunk 4 above)
Comment (brennj) [reply src/tests/jsonLogic.test.js]: Yeah, you're dead right - exact same thing. added small note in ee18568ba2edc9717fda8ae1c3e59058386f5f07

File: src/tests/jsonLogic.test.js (hunk 12 above)
Comment (johnstonbl01) [reply src/tests/jsonLogic.test.js:404]: Even after reading the comments and code, I'm not 100% sure on this one. We can proceed and then come back and adjust if needed?

**Review by johnstonbl01 — APPROVED** [review]

File: src/tests/jsonLogic.test.js (hunk 12 above)
Comment (sandrina-p) [reply src/tests/jsonLogic.test.js:404]: Looking at this from the outside, `NaN` doesn't make sense. When the json-logic is incomplete (aka missing a value for the calculation), I'd rather not run it and just return `undefined`, as if it didn't exist. WDYT?

Since this is an inoffensive edge-case (I hope!) and don't want to slow you down, I'd just leave a comment explaining why it's NaN and the linear issue ID for us to follow-up later on.

**Review by sandrina-p — APPROVED** [review]
